pkg/osutil/osutil.go

```diff
@@ -33,3 +37,49 @@ func Unsetenv(key string) error {
 	}
 	return nil
 }
+
+// InterruptHandler is a function that is called on receiving a
+// SIGTERM or SIGINT signal.
+type InterruptHandler func()
+
+var (
+	interruptMu sync.Mutex
+	// interruptHandlers holds all registered InterruptHandlers in order
+	// they will be executed.
+	interruptHandlers = []InterruptHandler{}
+)
+
+// RegisterInterruptHandler registers a new InterruptHandler.
+func RegisterInterruptHandler(h InterruptHandler) {
+	interruptMu.Lock()
+	defer interruptMu.Unlock()
+	interruptHandlers = append(interruptHandlers, h)
+}
+
+// HandleInterrupts calls the handler functions on receiving a SIGINT or SIGTERM.
+func HandleInterrupts() {
+	notifier := make(chan os.Signal, 1)
+	signal.Notify(notifier, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		sig := <-notifier
+		signal.Stop(notifier)
+
+		interruptMu.Lock()
+
+		log.Printf("received %v signal, shutting down\n", sig)
+
+		for _, h := range interruptHandlers {
+			h()
```

Comment:
If we want to make this pkg better, we probably should not call h() with the interruptLock. 

If the termination of h() depends on the interrupt lock, then there might be deadlock. For example, you are terminating a func that is trying to register an interrupt handler.